### ISSUE TITLE
Calling `uc_emu_stop` on a separate thread doesn't work if a hook is hanging.

### ISSUE THREAD

**Issue by macmade** [unicorn-engine/unicorn#1112]
I'm using the `UC_HOOK_CODE` hook for single-stepping emulated code.  
This basically implies waiting inside the hook for user input, before continuing.

Now based on some other condition, I want to be able to stop the emulation from a different thread, while the hook function is still waiting for user input.

In such a scenario, the emulation will never stop, although `uc_emu_stop` returns `UC_ERR_OK`.
  
There's probably a locked mutex while the hook is executing which prevents stopping.

I'm not sure if this could be by design, of if this is a bug.  
Anyway here's some code to demonstrate the issue:

    #include <unicorn/unicorn.h>
    #include <thread>
    #include <iostream>
    #include <array>

    static void onCode( uc_engine * uc, uint64_t address, uint32_t size, void * data );
    static void onCode( uc_engine * uc, uint64_t address, uint32_t size, void * data )
    {
        ( void )uc;
        ( void )address;
        ( void )size;
        ( void )data;
        
        std::cout << "Instruction break" << std::endl;
        
        while( 1 )
        {
            std::this_thread::yield();
        }
    }

    int main( void )
    {
        uc_engine * uc;
        uc_hook     h;
        
        uc_open( UC_ARCH_X86, UC_MODE_64, &uc );
        uc_mem_map( uc, 0, 1024 * 1024, UC_PROT_ALL );
        uc_hook_add( uc, &h, UC_HOOK_CODE, reinterpret_cast< void * >( onCode ), NULL, 0, 1024 * 1024 );
        
        {
            std::array< uint8_t, 6 > instr = { 0x48, 0xFF, 0xC1, 0x48, 0xFF, 0xC1 };
            
            uc_mem_write( uc, 0, instr.data(), instr.size() );
        }
        
        std::thread
        (
            [ & ]
            {
                std::this_thread::sleep_for( std::chrono::seconds( 1 ) );
                
                uc_err e;
                
                e = uc_emu_stop( uc );
                
                std::cout << uc_strerror( e ) << std::endl;
            }
        )
        .detach();
        
        uc_emu_start( uc, 0, 1024 * 1024, 0, 0 );
        
        return 0;
    }

**Comment by macmade** [comment]
Fair enough, thanks for the explanation 😊

**Comment by goodguysoft** [comment]
I have the same issue. Even a lot of text lines is written in previous post ;) to explain "how does it work", as it often appears with Unicorn, it doesn't work in such a way. `uc_emu_stop` is ignored silently inside code hook often.

I have the issue in single threaded program.

`uc_emu_stop` really does its job and sets `stop_request`, `cpu->exit_request` and `cpu->tcg_exit_req` to nonzero values, then it returns `UC_ERR_OK` without any real error checks.

After hook, the code returns to `cpu_tb_exec`. Here (`cpu-exec.c`, line 321) it clears `cpu->tcg_exit_req`.

Next, it clears also `cpu->exit_request` (`cpu-exec.c`, line 207) and calls `cpu_loop_exit` (line 209).

But the only visible effect of `cpu_loop_exit` is clearing of all "exit flags". The loop continues like a charm...

**Comment by gerph** [comment]
I'm sorry you didn't find that my comment was helpful. As I user of the library I was trying to give an explanation of what I know has worked for me and how, after examining the code, I understand it to work. I provided a number of explanations, feeling confident that my answers would be helpful and were informed about how things were. I am sorry that you don't feel that they were helpful.
